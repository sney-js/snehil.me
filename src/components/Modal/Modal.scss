.d-modal {
  background: var(--color-highlight-4);
  bottom: 0;
  display: none;
  left: 0;
  overflow: hidden;
  position: fixed;
  right: 0;
  top: 0;
  margin-top: var(--nav-height);
  z-index: var(--z-index-xs);
  animation: modal-fade-animation-out var(--transition);

  &.active {
    display: block;
    animation: modal-fade-animation var(--transition);
    background-color: var(--color-highlight);
  }

  &__close {
    width: 30px;
    height: 30px;
    position: fixed;
    display: inline-block;
    right: var(--spacing-gap);
    top: var(--spacing-gap);
    border: 2px var(--border-style) var(--color-primary-dynamic);
    color: var(--color-primary-dynamic);
    border-radius: 50%;
    background: none;
    padding: 6px;
    cursor: pointer;
  }

  &__dialog {
    background: var(--color-bg);
    color: var(--color-body);
    left: 50%;
    position: absolute;
    top: 50%;
    border-radius: 8px;
    overflow: hidden;
    padding: var(--gap-layout-tablet) var(--spacing-gap);
    width: auto;
    transform: translate(-50%, -50%);
    width: 80%;
    max-width: 800px;
    max-height: calc(100vh - var(--nav-height) - var(--spacing-gap-m));

    @media #{$screen-mobile} {
      width: calc(100vw - var(--spacing-gap) * 2);
    }
  }

  &__title {
    margin-top: var(--spacing-gap-xs);
    margin-bottom: var(--spacing-gap-xs);
    color: var(--color-body);
    padding-right: var(--spacing-gap-m);
  }

  &__body {
    overflow-y: auto;
    max-height: 80vh;
    padding: 0 var(--spacing-gap);
  }

  &__buttonGroup {
    margin-top: var(--spacing-gap);
    float: right;
    @media #{$screen-mobile} {
      float: none;
      margin-top: var(--spacing-gap-m);
    }
  }

  &__underlay {
    width: 100%;
    height: 100%;
  }
  
  &.active {
    &__underlay {
      background-color: var(--color-highlight);
    }
  }
}
