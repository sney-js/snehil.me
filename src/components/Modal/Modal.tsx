import React, { FC } from 'react';
import { makeClass } from 'utils/Helpers';
import Button from 'elements/Button/';
import { IcClose } from 'elements/SvgElements';
import Grid from 'components/Grid';
import { LinkType } from '../../models';
import Link from '../../elements/Link';

export type ModalProps = {
  /**
   * The main title that appears on the top
   */
  title?: string;
  /**
   * Determines the state of modal
   * e.g.
   * `true | false`
   */
  open?: boolean;
  /**
   * Determines the content of modal
   * e.g.
   */
  children?: JSX.Element | JSX.Element[];
  /**
   * Should be provided to have main CTA. `to` can be left undefined for a button onClick
   */
  primaryLink?: LinkType;
  /**
   * Should be provided to have secondary CTA. `to` can be left undefined for a button onClick
   */
  secondaryLink?: LinkType;
  /**
   * Event when closeButton is clicked
   */
  onClickClose?: () => void | undefined;
} & React.HTMLAttributes<HTMLDivElement>;

/**
 * Use `Modal` to highlight key info with a predefined status.
 */
const Modal: FC<ModalProps> = (props: ModalProps) => {
  const {
    open,
    children,
    className,
    onClickClose,
    title,
    primaryLink,
    secondaryLink
  } = props;
  const classes = makeClass(['d-modal', open && 'active', className]);

  return (
    <div className={classes}>
      <div className='d-modal__dialog'>
        <Button
          className='d-modal__close'
          onClick={() => onClickClose && onClickClose()}
          aria-label='Menu'
          appearance='secondary'
          icon={<IcClose />}
        />
        <div className='d-modal__body'>
          <h4 className='d-modal__title'>{title}</h4>
          <div className='d-modal__content'>{children}</div>
          <Grid
            className='d-modal__buttonGroup'
            template='auto auto'
            templateMobile='1fr'
          >
            {secondaryLink && (
              <Link
                {...secondaryLink}
                appearance={secondaryLink.appearance || 'secondary'}
              />
            )}
            {primaryLink && (
              <Link
                {...primaryLink}
                appearance={primaryLink.appearance || 'primary'}
              />
            )}
          </Grid>
        </div>
      </div>
      <div
        className='d-modal__underlay'
        onClick={() => onClickClose && onClickClose()}
      />
    </div>
  );
};

Modal.displayName = 'Modal';
Modal.defaultProps = {
  open: false
};

export default Modal;
