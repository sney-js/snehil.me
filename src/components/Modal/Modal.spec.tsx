import * as React from 'react';
import Modal from './Modal';
import { shallow } from 'enzyme';

const props = (
  <div>
    <div style={{ fontSize: 32 }}>About For Good Causes</div>
    <p>
      Choose to do something truly amazing with your BP points and donate them
      to a charity of your choice. Via For Good Causes, you can donate to a
      charity that close to your heart and help make a difference today! Over
      17,000 charities are available from global causes to small local
      charities, in the heart of your community. With charities covering every
      cause, including the environment, medical research, children and the
      young, animals and conservation, mental health and many more, there is
      something for everyone. Because we all care about different things. Once
      you have logged in, select the amount of points you want to exchange Click
      proceed and confirm your redemption. We will send you an email with your
      For Good Causes charity token To donate go to
      https://my.forgoodcauses.org/ BP Terms n Conditions apply, see
      https://www.bpmerewards.co.uk/terms-and-conditions for full details. For
      Good Causes Terms n Conditions apply, see
      https://www.forgoodcauses.org/terms-and-conditions/ for full details. For
      Good Causes deducts a small commission from each donation to cover the
      cost of the distribution to over 17,000 charities.
    </p>
  </div>
);
describe('<Modal />', () => {
  test('Rendered component with html as child', () => {
    const results = shallow(<Modal open={true}>{props}</Modal>);

    expect(results.exists()).toBe(true);
    expect(results.html()).toMatchSnapshot();
  });
  // FIXME
  // test('Should not rendered component if click close button', () => {
  //   const results = shallow(<Modal open={true}>{props}</Modal>);

  //   results.find('.d-modal__close').simulate('click');
  //   expect(results.find('.active').length).toBe(0);
  // });
});
