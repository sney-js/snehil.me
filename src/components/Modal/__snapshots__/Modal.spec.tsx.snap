// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> Rendered component with html as child 1`] = `"<div class=\\"d-modal active\\"><div class=\\"d-modal__dialog\\"><button class=\\"d-modal__close d-button secondary icon-primary\\" aria-label=\\"Menu\\"><i class=\\"d-button-icon\\"><svg viewBox=\\"0 0 24 24\\" fill=\\"currentColor\\"><path d=\\"M6.225 4.811a1 1 0 00-1.414 1.414L10.586 12 4.81 17.775a1 1 0 101.414 1.414L12 13.414l5.775 5.775a1 1 0 001.414-1.414L13.414 12l5.775-5.775a1 1 0 00-1.414-1.414L12 10.586 6.225 4.81z\\" fill=\\"currentColor\\"></path></svg></i></button><div class=\\"d-modal__body\\"><h4 class=\\"d-modal__title\\"></h4><div class=\\"d-modal__content\\"><div><div style=\\"font-size:32px\\">About For Good Causes</div><p>Choose to do something truly amazing with your BP points and donate them to a charity of your choice. Via For Good Causes, you can donate to a charity that close to your heart and help make a difference today! Over 17,000 charities are available from global causes to small local charities, in the heart of your community. With charities covering every cause, including the environment, medical research, children and the young, animals and conservation, mental health and many more, there is something for everyone. Because we all care about different things. Once you have logged in, select the amount of points you want to exchange Click proceed and confirm your redemption. We will send you an email with your For Good Causes charity token To donate go to https://my.forgoodcauses.org/ BP Terms n Conditions apply, see https://www.bpmerewards.co.uk/terms-and-conditions for full details. For Good Causes Terms n Conditions apply, see https://www.forgoodcauses.org/terms-and-conditions/ for full details. For Good Causes deducts a small commission from each donation to cover the cost of the distribution to over 17,000 charities.</p></div></div><section class=\\"d-modal__buttonGroup layout-grid\\" style=\\"--grid-template:auto auto;--grid-template-mobile:1fr;--grid-gap:var(--spacing-gap);--grid-align:center\\"></section></div></div></div>"`;
